sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XXE/XmlReaderSettingsValidator.cs

```diff
@@ -0,0 +1,106 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2020 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.XXE
+{
+    internal class XmlReaderSettingsValidator
+    {
+        private readonly SemanticModel semanticModel;
+
+        public XmlReaderSettingsValidator(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        public bool IsUnsafe(InvocationExpressionSyntax invocation, ISymbol settings, NetFrameworkVersion version)
+        {
+            var unsafeDtdProcessing = false;
+            var unsafeResolver = version == NetFrameworkVersion.Probably35 || version == NetFrameworkVersion.Between4And451;
+
+            var objectCreation = GetObjectCreation(settings, invocation, semanticModel);
+            var objectCreationAssignments = objectCreation.GetInitializerExpressions().OfType<AssignmentExpressionSyntax>();
+
+            var assignments = invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>()
+                .DescendantNodes()
+                .OfType<AssignmentExpressionSyntax>()
+                .Where(assignment => assignment.Left is MemberAccessExpressionSyntax memberAccess &&
+                                     IsXmlReaderSettings(memberAccess.Expression, semanticModel) &&
+                                     settings.Equals(semanticModel.GetSymbolInfo(memberAccess.Expression).Symbol));
+
+            foreach (var assignment in objectCreationAssignments.Union(assignments))
+            {
+                if (assignment.Left.NameIs("ProhibitDtd") ||
```

Comment:
you could extract `assignment.Left.Name` (whatever the name is) in a local variable and just do string comparisons to avoid calling a method each time